src/backends/github/API.js

```diff
@@ -323,42 +333,158 @@ export default class API {
+     * return commits as is.
+     */
+    if (commits.length === 0 || commits[0].parents[0].sha === baseCommit.sha) {
+      return Promise.resolve(last(commits));
     }
+
+    /**
+     * Re-create each commit over the new base, applying each to the previous,
+     * changing only the parent SHA and tree for each, but retaining all other
+     * info, such as the author/committer data.
+     */
+    const newHeadPromise = commits.reduce((lastCommitPromise, commit, idx) => {
+      return lastCommitPromise.then(newParent => {
+        /**
+         * Use `baseCommit` as the parent of the first commit, and normalize
+         * commit data to ensure it's not nested in `commit.commit`.
+         */
+        const parent = this.normalizeCommit(idx === 0 ? baseCommit : newParent);
+        const commitToRebase = this.normalizeCommit(commit);
+
+        return this.rebaseSingleBlobCommit(parent, commitToRebase, pathToBlob);
+      });
+    }, Promise.resolve());
+
+    /**
+     * Return a promise that resolves when all commits have been created.
+     */
+    return newHeadPromise;
+  }
+
+  /**
+   * Rebase a commit that changes a single blob. Also handles updating the tree.
+   */
+  rebaseSingleBlobCommit(baseCommit, commit, pathToBlob) {
+    /**
+     * Retain original commit metadata.
+     */
+    const { message, author, committer } = commit;
+
+    /**
+     * Set the base commit as the parent.
+     */
+    const parent = [ baseCommit.sha ];
+
+    /**
+     * Get the blob data by path.
+     */
+    return this.getBlobInTree(commit.tree.sha, pathToBlob)
+
+      /**
+       * Create a new tree consisting of the base tree and the single updated
+       * blob. Use the full path to indicate nesting, GitHub will take care of
+       * subtree creation.
+       */
+      .then(blob => this.createTree(baseCommit.tree.sha, [{ ...blob, path: pathToBlob }]))
+
+      /**
+       * Create a new commit with the updated tree and original commit metadata.
+       */
+      .then(tree => this.createCommit(message, tree.sha, parent, author, committer));
+  }
+
+
+  /**
+   * Get a pull request by PR number.
+   */
+  getPullRequest(prNumber) {
+    return this.request(`${ this.repoURL }/pulls/${prNumber} }`);
+  }
+
+  /**
+   * Get the list of commits for a given pull request. Optionally assert a head
+   * commit.
+   */
+  getPullRequestCommits(prNumber, assertHead) {
+    return this.request(`${ this.repoURL }/pulls/${prNumber}/commits`).then(prCommits => {
+      /**
+       * Sometimes the list of commits for a pull request isn't updated
+       * immediately after the PR branch is patched. If the branch is known to
+       * have been updated very recently, `assertHead` will be appended, but
+       * only if the last commit in the response is the parent of `assertHead`.
+       */
+      const missingHead = assertHead && assertHead.parents[0].sha === last(prCommits).sha;
```

Comment:
To start off with, __I'm fine merging this PR with this code as is__ since it's not breaking anything, I just want to bring this up for discussion. `assertHead` implies to me that the function will error if I pass `assertHead` and neither it nor its parent(s) matches the most recent SHA. Instead, `getPullRequestCommits`  silently continues, which seems prone to mishap - if the repo moves out from under us we probably don't want to blindly continue our mutations. OTOH, I can see people using `assertHead` without making mutations, who don't _want_ an error here. A good fix IMO would be to separate this into two different methods - `getPullRequestCommits` (with no `assertHead` parameter) and `assertHead` (which does the conditional concatenation, and can have its errors caught and handled separately from API or HTTP errors)  - which would allow the caller to decide how to handle it on a case-by-case basis, and simplify the related functions. (the equivalent to what this does now would be `getPullRequestCommits(prNumber).then(commits => assertHead(commits, head).catch(() => commits))`, which IMO expresses the intent much more clearly, eliminates a confusing argument from `getPullRequestCommits`, and makes what should be the most common desired behavior of `assertHead` (erroring when mutating a repo which has state we didn't expect) the default.

Reply:
Yep, I'm with that.